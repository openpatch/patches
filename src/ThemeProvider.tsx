import React, { ReactElement } from "react";
import { ThemeProvider as EThemeProvider } from "emotion-theming";
import { Theme, base } from "./themes";
import { GlobalCssTheme } from "./themes/utils";

interface ThemeProviderProps {
  theme?: Theme;
  children: ReactElement;
}

const ThemeProvider = ({ theme = base, children }: ThemeProviderProps) => {
  return (
    <EThemeProvider theme={theme}>
      <GlobalCssTheme theme={theme} />
      {children}
    </EThemeProvider>
  );
};

export default ThemeProvider;
