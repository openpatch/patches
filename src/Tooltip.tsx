import { css } from "@emotion/react";
import { ReactNode, useEffect, useRef, useState } from "react";
import { Box } from "./Box";

export type TooltipProps = {
  children: ReactNode;
  tooltip: ReactNode;
  width?: number;
  position?: "top" | "left" | "right" | "bottom";
};

export const Tooltip = ({
  children,
  tooltip,
  position = "top",
  width = 120,
}: TooltipProps) => {
  const tooltipRef = useRef<HTMLSpanElement>(null);
  const [height, setHeight] = useState(0);
  useEffect(() => {
    if (tooltipRef.current) {
      setHeight(tooltipRef.current.clientHeight);
    }
  }, []);

  return (
    <Box
      css={css`
        position: relative;
        display: inline-block;

        &:hover .tooltip {
          opacity: 1;
        }
      `}
    >
      {children}
      <Box
        as="span"
        ref={tooltipRef as any}
        className="tooltip"
        css={(theme) => [
          css`
            width: ${width}px;
            background-color: black;
            color: #fff;
            text-align: center;
            padding: ${theme.space.xsmall} ${theme.space.small};
            border-radius: ${theme.radii.standard};
            opacity: 0;
            transition: opacity 0.3s;

            /* Position the tooltip text - see examples below! */
            position: absolute;
            z-index: 1;
          `,
          position === "bottom" &&
            css`
              top: 150%;
              left: 50%;
              margin-left: -${width / 2}px;

              &::after {
                content: " ";
                position: absolute;
                bottom: 100%; /* At the top of the tooltip */
                left: 50%;
                margin-left: -5px;
                border-width: 5px;
                border-style: solid;
                border-color: transparent transparent black transparent;
              }
            `,
          position === "right" &&
            css`
              top: -${height / 2 - 10}px;
              left: 110%;

              &::after {
                content: " ";
                position: absolute;
                top: 50%;
                right: 100%; /* To the left of the tooltip */
                margin-top: -5px;
                border-width: 5px;
                border-style: solid;
                border-color: transparent black transparent transparent;
              }
            `,
          position === "left" &&
            css`
              top: -${height / 2 - 10}px;
              right: 110%;
              &::after {
                content: " ";
                position: absolute;
                top: 50%;
                left: 100%;
                margin-top: -5px;
                border-width: 5px;
                border-style: solid;
                border-color: transparent transparent transparent black;
              }
            `,
          position === "top" &&
            css`
              bottom: 150%;
              left: 50%;
              margin-left: -${width / 2}px;
              &::after {
                content: " ";
                position: absolute;
                top: 100%; /* At the bottom of the tooltip */
                left: 50%;
                margin-left: -5px;
                border-width: 5px;
                border-style: solid;
                border-color: black transparent transparent transparent;
              }
            `,

          height === 0 &&
            css`
              opacity: 0;
            `,
        ]}
      >
        {tooltip}
      </Box>
    </Box>
  );
};
