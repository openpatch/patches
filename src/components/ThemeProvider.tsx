import React, { ReactElement } from "react";
import {
  ThemeProvider as SCThemeProvider,
  DefaultTheme,
} from "styled-components";
import defaultTheme from "../shared/theme";

interface ThemeProviderProps {
  theme?: DefaultTheme;
  children: ReactElement;
}

const ThemeProvider = ({
  theme = defaultTheme,
  children,
}: ThemeProviderProps) => {
  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};

export default ThemeProvider;
