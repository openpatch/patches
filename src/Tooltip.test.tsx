import { render } from "@testing-library/react";
import { Tooltip, TooltipProps } from "./Tooltip";
import { ThemeProvider } from "./ThemeProvider";

const renderWithTheme = (props?: TooltipProps) =>
  render(
    <ThemeProvider>
      <Tooltip {...props} />
    </ThemeProvider>
  );

test("should render", () => {
  renderWithTheme();
});

test("should match snapshot", () => {
  const { container } = renderWithTheme();
  expect(container).toMatchSnapshot();
});
