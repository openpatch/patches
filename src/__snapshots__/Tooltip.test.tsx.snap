// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
.emotion-0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  position: relative;
  display: inline-block;
}

.emotion-0:hover .tooltip {
  opacity: 1;
}

.emotion-2 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  width: 120px;
  background-color: black;
  color: #fff;
  text-align: center;
  padding: 8px 16px;
  border-radius: 8px;
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
  position: absolute;
  z-index: 1;
  bottom: 150%;
  left: 50%;
  margin-left: -60px;
  opacity: 0;
}

.emotion-2::after {
  content: " ";
  position: absolute;
  top: 100%;
  left: 50%;
  margin-left: -5px;
  border-width: 5px;
  border-style: solid;
  border-color: black transparent transparent transparent;
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <span
      class="tooltip emotion-2 emotion-1"
    />
  </div>
</div>
`;
